### PR TITLE
ui-materialui: Add support for `<Edit emptyWhileLoading>`

### PR DESCRIPTION
**Pull request by erwanMarmelab**
## Problem

The Show and List components support that prop, not the Edit.

## Solution

Add support for `<Edit emptyWhileLoading>`

## Additional Checks

- [x] The PR targets `master` for a bugfix, or `next` for a feature
~- [ ] The PR includes **unit tests** (if not possible, describe why)~
- [x] The PR includes one or several **stories** (if not possible, describe why)
- [x] The **documentation** is up to date

### CODE REVIEW

File: packages/ra-ui-materialui/src/detail/Edit.stories.tsx [hunk 1]
```diff
@@ -300,3 +300,45 @@ export const Default = () => (
         </Admin>
     </TestMemoryRouter>
 );
+
+export const emptyWhileLoading = () => {
```
Comment (djhi) [inline packages/ra-ui-materialui/src/detail/Edit.stories.tsx:304]:
```suggestion
export const EmptyWhileLoading = () => {
```

**Review by djhi — CHANGES_REQUESTED** [review]
Can you also add a test?

File: docs/Edit.md [hunk 2]
```diff
@@ -222,6 +222,57 @@ const PostEdit = () => (
 );
 ```
 
+## `emptyWhileLoading`
+
+By default, `<Edit>` renders its child component even before the `dataProvider.getOne()` call returns. And default layout components (`<Datagrid>` and `<SimpleList>`) return null when the data is loading. If you use a custom layout component instead, you'll have to handle the case where the `data` is not yet defined.
```
Comment (fzaninotto) [inline docs/Edit.md:227]: Why mention `Datagrid` and `SimpleList` here? We're not in a list page.

File: docs/Edit.md [hunk 3]
```diff
@@ -222,6 +222,57 @@ const PostEdit = () => (
...
+
+But if you use a custom child component that expects the record context to be defined, your component will throw an error. For instance, the following will fail on load with a "ReferenceError: data is not defined" error:
```
Comment (fzaninotto) [inline docs/Edit.md:229]: The first sentence says more or less the same as the previous one... not clear. The second mentions an error about undefined data but there is no data in the snippet.

File: docs/Edit.md [hunk 4]
```diff
@@ -222,6 +222,57 @@ const PostEdit = () => (
 );
 ```
 
+## `emptyWhileLoading`
```
Comment (fzaninotto) [inline docs/Edit.md:225]: don't forget to add it to the list of props earlier in the doc

File: docs/EditTutorial.md [hunk 5]
```diff
@@ -479,7 +479,7 @@ import * as React from "react";
 import { Edit, SimpleForm, TextInput, SelectInput } from "react-admin";
 
 export const BookEdit = () => (
-  <Edit>
+  <Edit emptyWhileLoading>
```
Comment (fzaninotto) [inline docs/EditTutorial.md:482]: I think you should revert the changes in this file. We don't need to introduce this prop, which is for corner case, in a general presentation of the Edit logic.

File: packages/ra-ui-materialui/src/detail/Edit.spec.tsx [hunk 6]
```diff
@@ -106,6 +107,13 @@ describe('<Edit />', () => {
         });
     });
 
+    it("shoudln't display the record while loading", async () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/Edit.spec.tsx:110]:
```suggestion
    it("shoudln't display the Edit child while loading when emptyWhileLoading is true", async () => {
```

File: packages/ra-ui-materialui/src/detail/Edit.spec.tsx [hunk 7]
```diff
@@ -106,6 +107,13 @@ describe('<Edit />', () => {
...
+        render(<EmptyWhileLoading />);
+        expect(screen.queryByText('Book War and Peace')).toBeNull();
+        await new Promise(resolve => setTimeout(resolve, 3000));
+        expect(screen.queryByText('Book War and Peace')).not.toBeNull();
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/Edit.spec.tsx:114]: This test doesn't make sense. Of course the record won't show up during loading, because the data isn't there. The purpose of the `emptyWhileLoading` prop is to avoid rendering the child altogether. 
So you should add a static string to the Edit child and make sure it's not rendered while the data provider is loading. 

Also, don't use timeouts. Check https://github.com/marmelab/react-admin/pull/10238/files#diff-1f5141788db3c747028c3cb2da9eaa6f9244e63d77addb0ea7eb70688fcb9ba9 for an example of test that uses a suspended Promise resolved bu the test itself.

File: packages/ra-ui-materialui/src/detail/EditView.tsx [hunk 8]
```diff
@@ -25,7 +26,7 @@ export const EditView = (props: EditViewProps) => {
 
     const finalActions =
         typeof actions === 'undefined' && hasShow ? defaultActions : actions;
-    if (!children) {
+    if (!children || (!record && emptyWhileLoading)) {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/EditView.tsx:29]: you must also check the `pending` state, as in the other controllers.

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: docs/Edit.md [hunk 9]
```diff
@@ -222,6 +223,57 @@ const PostEdit = () => (
...
+    if (isPending) return null;
+    return (
+        <Typography>
+            <i>{record.title}</i>, by {record.author} ({record.year})
+        </Typography>
+    );
+}
+```
+
+The `<Edit emptyWhileLoading>` prop provides a convenient shortcut for that use case. When enabled, `<Edit>` won't render its child until `data` is defined.
```
Comment (erwanMarmelab) [inline docs/Edit.md:266]:
```suggestion
The `<Edit emptyWhileLoading>` prop provides a convenient shortcut for that use case. When enabled, `<Edit>` won't render its `aside`, `actions` and `title` components until `record` is defined.
```

File: docs/EditTutorial.md [hunk 10]
```diff
@@ -495,4 +495,7 @@ export const BookEdit = () => (
 
 React-admin components are not magic, they are React components designed to let you focus on the business logic and avoid repetitive tasks. 
 
-Tip: Actually, `<Edit>` does more than the code it replaces in the previous example: it handles notification and redirection upon submission, it sets the page title, and handles the error logic.
+**Tip:** Actually, `<Edit>` does more than the code it replaces in the previous example: it handles notification and redirection upon submission, it sets the page title, and handles the error logic.
+
+**Tip**: With `emptyWhileLoading` turned on, the `<Edit>` component doesn't render its child component until the record is available. Without this flag, the Field components would render even during the loading phase, and may break if they aren't planned to work with an empty record context.
```
Comment (erwanMarmelab) [inline docs/EditTutorial.md:500]:
```suggestion
**Tip**: With `emptyWhileLoading` turned on, the `<Edit>` component doesn't render its `aside`, `actions` and `title` components until the record is available. Without this flag, the Field components would render even during the loading phase, and may break if they aren't planned to work with an empty record context.
```

File: docs/Edit.md [hunk 11]
```diff
@@ -222,6 +223,57 @@ const PostEdit = () => (
 );
 ```
 
+## `emptyWhileLoading`
+
+By default, `<Edit>` doesn't render its child component even before the `dataProvider.getOne()` call returns, but it returns its [`aside`](#aside), [`actions`](#actions) and [`title`](#title) components.
+
+If you use one of those components customized with a record, your component will throw an error. For instance, the following will fail on load with a "ReferenceError: record is not defined" error:
```
Comment (djhi) [inline docs/Edit.md:230]:
```suggestion
If a component you use in those props relies on the record, it will throw an error. For instance, the following will fail on load with a "ReferenceError: record is not defined" error:
```

**Review by djhi — CHANGES_REQUESTED** [review]

File: docs/EditTutorial.md [hunk 12]
```diff
@@ -495,4 +495,7 @@ export const BookEdit = () => (
 
 React-admin components are not magic, they are React components designed to let you focus on the business logic and avoid repetitive tasks. 
 
-Tip: Actually, `<Edit>` does more than the code it replaces in the previous example: it handles notification and redirection upon submission, it sets the page title, and handles the error logic.
+**Tip:** Actually, `<Edit>` does more than the code it replaces in the previous example: it handles notification and redirection upon submission, it sets the page title, and handles the error logic.
+
+**Tip**: With `emptyWhileLoading` turned on, the `<Edit>` component doesn't render its `aside`, `actions` and `title` components until the record is available. Without this flag, those components would render even during the loading phase, and may break if they aren't planned to work with an empty record context.
```
Comment (fzaninotto) [inline docs/EditTutorial.md:500]: This is advanced configuration and we shouldn't mention it in the EditTutorial, which is a general introduction.

File: packages/ra-ui-materialui/src/detail/Edit.spec.tsx [hunk 13]
```diff
@@ -106,6 +107,27 @@ describe('<Edit />', () => {
...
+        expect(screen.queryByText('Something went wrong')).toBeNull();
+
+        await waitFor(
+            () => {
+                expect(
+                    screen.queryByText('War and Peace, by Leo Tolstoy (1869)')
+                ).not.toBeNull();
+            },
+            { timeout: 2500 }
+        );
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/Edit.spec.tsx:126]: This will slow down the unit tests. Instead of relying on a timeout, inject a data provider with a suspended Promise that you resolve manually.

File: packages/ra-ui-materialui/src/detail/Edit.spec.tsx [hunk 14]
```diff
@@ -885,7 +907,12 @@ describe('<Edit />', () => {
                     i18nProvider={i18nProvider}
                 >
                     <ResourceDefinitionContextProvider
-                        definitions={{ foo: { recordRepresentation: 'title' } }}
+                        definitions={{
+                            foo: {
+                                recordRepresentation: 'title',
+                                name: '',
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/Edit.spec.tsx:913]: does not make sense

File: packages/ra-ui-materialui/src/detail/Edit.spec.tsx [hunk 15]
```diff
@@ -885,7 +919,9 @@ describe('<Edit />', () => {
                     i18nProvider={i18nProvider}
                 >
                     <ResourceDefinitionContextProvider
-                        definitions={{ foo: { recordRepresentation: 'title' } }}
+                        definitions={{
+                            foo: { recordRepresentation: 'title' },
```
Comment (djhi) [inline packages/ra-ui-materialui/src/detail/Edit.spec.tsx:923]:
```suggestion
                            foo: { name: 'foo', recordRepresentation: 'title' },
```

I believe TS is angry about this

**Review by djhi — COMMENTED** [review]
Would you mind fixing the Show documentation while you're at it? The `emptyWhileLoading` description in the props table is wrong:
> Set to true to return null while the list is loading

**Review by fzaninotto — APPROVED** [review]
